frontend/src/metabase/components/TextWidget/TextWidget.unit.spec.tsx

```diff
@@ -10,56 +11,83 @@ const TextInputWithStateWrapper = ({ value }: { value?: number | string }) => {
   );
 };
 
+const PLACEHOLDER_TEXT = "Enter a value...";
+
 describe("TextWidget", () => {
   it("should render correctly", () => {
     render(
       <TextWidget
-        value={"Hello, world!"}
+        value="Hello, world!"
         setValue={jest.fn()}
         focusChanged={jest.fn()}
-      ></TextWidget>,
+      />,
     );
-    expect(screen.getByRole("textbox")).toHaveValue("Hello, world!");
+
+    const textbox = screen.getByPlaceholderText(PLACEHOLDER_TEXT);
+
+    expect(textbox).toHaveValue("Hello, world!");
   });
 
   it("should accept editing", () => {
     render(
-      <TextWidget
-        value={""}
-        setValue={jest.fn()}
-        focusChanged={jest.fn()}
-      ></TextWidget>,
+      <TextWidget value={""} setValue={jest.fn()} focusChanged={jest.fn()} />,
     );
 
-    fireEvent.change(screen.getByRole("textbox"), {
-      target: { value: "Toucan McBird" },
-    });
-    expect(screen.getByRole("textbox")).toHaveValue("Toucan McBird");
+    const textbox = screen.getByPlaceholderText(PLACEHOLDER_TEXT);
+
+    userEvent.paste(textbox, "Toucan McBird");
+
+    expect(textbox).toHaveValue("Toucan McBird");
   });
 
   it("should render a zero as an initial value", () => {
     render(
       <TextWidget value={0} setValue={jest.fn()} focusChanged={jest.fn()} />,
     );
 
-    expect(screen.getByRole("textbox")).toHaveValue("0");
+    const textbox = screen.getByPlaceholderText(PLACEHOLDER_TEXT);
+
+    expect(textbox).toHaveValue("0");
   });
 
-  it("should accept zero as an input value", async () => {
+  it("should accept zero as an input value", () => {
     render(<TextInputWithStateWrapper />);
 
-    const textbox = screen.getByRole("textbox");
+    const textbox = screen.getByPlaceholderText(PLACEHOLDER_TEXT);
+    userEvent.type(textbox, "0");
 
-    await userEvent.type(textbox, "0");
     expect(textbox).toHaveValue("0");
   });
 
-  it("should keep zero value when pressing enter", async () => {
+  it("should keep zero value when pressing enter", () => {
     render(<TextInputWithStateWrapper />);
 
-    const textbox = screen.getByRole("textbox");
+    const textbox = screen.getByPlaceholderText(PLACEHOLDER_TEXT);
+    userEvent.type(textbox, "0{enter}");
 
-    await userEvent.type(textbox, "0{enter}");
     expect(textbox).toHaveValue("0");
   });
+
+  it("should trigger setValue with null on blur if value is empty", () => {
+    const setValueSpy = jest.fn();
+
+    render(
+      <TextWidget value={""} setValue={setValueSpy} focusChanged={jest.fn()} />,
+    );
+
+    const textbox = screen.getByPlaceholderText(PLACEHOLDER_TEXT);
+    const textToType = "123";
+
+    userEvent.type(textbox, textToType);
+
+    _.times(textToType.length, () => {
+      userEvent.type(textbox, "{backspace}");
+    });
```

Comment:
`setValue` prop (which BTW should have been called `onChange`) in the `TextWidget` won't be called here so these 4 lines do not really affect the test.

Can we maybe call `userEvent.tab()` after `userEvent.type(textbox, textToType)`?

Reply:
but we cover exactly the case when we place something to the TextWidget and remove it afterwards. I'm sorry I didn't get your point

Reply:
Sorry, my point is that these 4 lines could be replaced with:

```
  textbox.focus();
```

`userEvent.type` will focus the input, so the test works.
But whether we're typing anything or removing anything does not actually matter.

Reply:
yes, but I wanted to simulate user behaviour anyway, so that was the plan